core/nakamoto/sync.go

```diff
@@ -0,0 +1,320 @@
+package nakamoto
+
+import (
+	"math/big"
+	"sync"
+	"time"
+
+	"github.com/liamzebedee/tinychain-go/core"
+)
+
+// Downloads block headers in parallel from a set of peers for a set of heights, relative to a base blockhash and height.
+//
+// The total number of headers we are downloading is represented by the count of items inside the heightMap.
+// The header size is estimated as 200 B. So for 2048 headers, we are downloading 409 KB.
+// The total download is then divided into a fixed-size workload of 50 KB each, which we call chunks.
+// These download chunk work items are then distributed to the peers in a round-robin fashion (ie. modulo).
+//
+// This function supports downloading as few as 1 header, which will download from a single peer, or 2048 headers, which
+// will download from as many as 9 peers in parallel.
+func (n *Node) downloadHeaders(fromNode [32]byte, fromHeight uint64, heightMap core.Bitstring, peers []Peer) []BlockHeader {
+	// Size of a block header is 200 B.
+	HEADER_SIZE := 200
+	
+	// Size of a chunk we request from a peer is 50 KB.
+	CHUNK_SIZE := 50*1000
+
+	// Total number of headers we're requesting.
+	NUM_HEADERS := heightMap.Count()
+
+	// Total number of chunks (download work items).
+	NUM_CHUNKS := (NUM_HEADERS * HEADER_SIZE / CHUNK_SIZE) + 1
+
+	// So for example:
+	// header_size = 200 B
+	// chunk_size = 50 KB
+	// ...
+		// num_headers = 100
+		// num_chunks = (100 * 200 / 50,000) + 1 = 1 chunks
+	// ...
+		// num_headers = 2048
+		// num_chunks = (2048 * 200 / 50,000) + 1 = 9 chunks
+
+	// Then we distribute these work items to our peers.
+	type ChunkWorkItem struct {
+		heights core.Bitstring
+	}
+	resultsChan := make(chan []BlockHeader, NUM_CHUNKS)
+	workItems := make([]ChunkWorkItem, NUM_CHUNKS)
+	
+	// Distribute the work:
+	// ...
+		// num_headers = 2048
+		// num_chunks = (2048 * 200 / 50,000) + 1 = 9 chunks
+		// work:
+			// 1: heights 0..79
+			// 2: heights 80..159
+			// N: heights 
+				// i * NUM_HEADERS / NUM_CHUNKS = i * 2048 / 9 = i * 227
+				// i*227 = 0, 227, 454, 681, 908, 1135, 1362, 1589, 1816
+	for i := 0; i < NUM_CHUNKS; i++ {
+		startHeight := i * (NUM_HEADERS / NUM_CHUNKS)
+		endHeight := (i + 1) * (NUM_HEADERS / NUM_CHUNKS)
+		heights := core.NewBitstring(heightMap.Size())
+		for j := startHeight; j < endHeight; j++ {
+			heights.SetBit(j)
+		}
+		workItems[i] = ChunkWorkItem{heights: *heights}
+	}
+
+	// Distribute the work items to our peers.
+	// for i, item := range workItems {
+	// 	peer := peers[i % len(peers)]
```

Comment:
TODO lol this will start 1000's of goroutines, should buffer one download per peer at a time